-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43219/37589

ERROR: Build errors found during clang-tidy run.
```
DataFormats/PortableTestObjects/src/classes.cc:4:72: error: expected ';' after top level declarator [clang-diagnostic-error]
    4 | SET_PORTABLEHOSTCOLLECTION_READ_RULES(portabletest::TestHostCollection)
      |                                                                        ^
      |                                                                        ;
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

--- a/DataFormats/Portable/README.md
+++ b/DataFormats/Portable/README.md
@@ -4,8 +4,37 @@
 
 `PortableHostCollection<T>` is a class template that wraps a SoA type `T` and an alpaka host buffer, which owns the
 memory where the SoA is allocated. The content of the SoA is persistent, while the buffer itself is transient.
-Specialisations of this template can be persisted, and can be read back also in "bare ROOT" mode, without any
-dictionaries.
+Specialisations of this template can be persisted, but requre specil ROOT read rules to read the data into a single
+memory buffer.
+
+The original way to declare these rules, now deprecated, is to add them to the `classes_def.xml` file. For example,
+to declare the read rules for `portabletest::TestHostCollection` based on the `portabletest::TestSoA` SoA, one
+would add to the same `classes_def.xml` file where `portabletest::TestHostCollection` is declared:
+```xml
+<read
+  sourceClass="portabletest::TestHostCollection"
+  targetClass="portabletest::TestHostCollection"
+  version="[1-]"
+  source="portabletest::TestSoA layout_;"
+  target="buffer_,layout_,view_"
+  embed="false">
+<![CDATA[
+  portabletest::TestHostCollection::ROOTReadStreamer(newObj, onfile.layout_);
+]]>
+```
+
+The new, recommended way to declare these rules is to create a `classes.cc` file along with `classes.h`, and use the
+`SET_PORTABLEHOSTCOLLECTION_READ_RULES`. For example, to declare the rules for `portabletest::TestHostCollection` one
+would create the file `classes.cc` with the content:
+```c++
+#include "DataFormats/Portable/interface/PortableHostCollectionReadRules.h"
+#include "DataFormats/PortableTestObjects/interface/TestHostCollection.h"
+
+SET_PORTABLEHOSTCOLLECTION_READ_RULES(portabletest::TestHostCollection);
+```
+
+`PortableHostCollection<T>` can also be read back in "bare ROOT" mode, without any dictionaries.
+
 They have no implicit or explicit references to alpaka (neither as part of the class signature nor as part of its name).
 This could make it possible to read them back with different portability solutions in the future.
 

--- a/DataFormats/PortableTestObjects/src/classes.cc
+++ b/DataFormats/PortableTestObjects/src/classes.cc
@@ -1,0 +1,4 @@
+#include "DataFormats/Portable/interface/PortableHostCollectionReadRules.h"
+#include "DataFormats/PortableTestObjects/interface/TestHostCollection.h"
+
+SET_PORTABLEHOSTCOLLECTION_READ_RULES(portabletest::TestHostCollection);

--- a/DataFormats/PortableTestObjects/src/classes_def.xml
+++ b/DataFormats/PortableTestObjects/src/classes_def.xml
@@ -1,18 +1,6 @@
 <lcgdict>
   <class name="portabletest::TestSoA"/>
   <class name="portabletest::TestSoA::View"/>
-
   <class name="portabletest::TestHostCollection"/>
-  <read
-    sourceClass="portabletest::TestHostCollection"
-    targetClass="portabletest::TestHostCollection"
-    version="[1-]"
-    source="portabletest::TestSoA layout_;"
-    target="buffer_,layout_,view_"
-    embed="false">
-  <![CDATA[
-    portabletest::TestHostCollection::ROOTReadStreamer(newObj, onfile.layout_);
-  ]]>
-  </read>
   <class name="edm::Wrapper<portabletest::TestHostCollection>" splitLevel="0"/>
 </lcgdict>

